src/dbnode/network/server/tchannelthrift/node/service.go

```diff
@@ -729,153 +728,277 @@ func (s *service) FetchTagged(tctx thrift.Context, req *rpc.FetchTaggedRequest)
 		)
 	}
 
-	result, err := s.fetchTagged(ctx, db, req)
+	result, err := s.fetchTagged(ctx, req)
 	if sampled && err != nil {
 		sp.LogFields(opentracinglog.Error(err))
 	}
 	sp.Finish()
 
+	if err != nil {
+		s.metrics.fetchTagged.ReportError(s.nowFn().Sub(callStart))
+	} else {
+		s.metrics.fetchTagged.ReportSuccess(s.nowFn().Sub(callStart))
+	}
+
 	return result, err
 }
 
-func (s *service) fetchTagged(ctx context.Context, db storage.Database, req *rpc.FetchTaggedRequest) (*rpc.FetchTaggedResult_, error) {
-	callStart := s.nowFn()
+func (s *service) fetchTagged(ctx context.Context, req *rpc.FetchTaggedRequest) (*rpc.FetchTaggedResult_, error) {
+	iter, err := s.FetchTaggedIter(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	var response rpc.FetchTaggedResult_
+	response.Elements = make([]*rpc.FetchTaggedIDResult_, 0, iter.NumIDs())
+	response.Exhaustive = iter.Exhaustive()
+	for iter.HasNextIDIter() {
+		idIter, err := iter.NextIDIter(ctx)
+		if err != nil {
+			return nil, err
+		}
+		id, err := idIter.IDResult()
+		if err != nil {
+			return nil, err
+		}
+		response.Elements = append(response.Elements, id)
+		for idIter.HasNextSegments() {
+			segments, err := idIter.NextSegments(ctx)
+			if err != nil {
+				return nil, err
+			}
+			id.Segments = append(id.Segments, segments)
+		}
+	}
+
+	return &response, nil
+}
+
+func (s *service) FetchTaggedIter(ctx context.Context, req *rpc.FetchTaggedRequest) (FetchTaggedResultsIter, error) {
+	db, err := s.startReadRPCWithDB()
+	if err != nil {
+		return nil, err
+	}
+	ctx.RegisterCloser(xresource.SimpleCloserFn(func() {
+		s.readRPCCompleted()
+	}))
 
 	ns, query, opts, fetchData, err := convert.FromRPCFetchTaggedRequest(req, s.pools)
 	if err != nil {
-		s.metrics.fetchTagged.ReportError(s.nowFn().Sub(callStart))
 		return nil, tterrors.NewBadRequestError(err)
 	}
 
 	queryResult, err := db.QueryIDs(ctx, ns, query, opts)
 	if err != nil {
-		s.metrics.fetchTagged.ReportError(s.nowFn().Sub(callStart))
 		return nil, convert.ToRPCError(err)
 	}
 
-	results := queryResult.Results
-	response := &rpc.FetchTaggedResult_{
-		Exhaustive: queryResult.Exhaustive,
-		Elements:   make([]*rpc.FetchTaggedIDResult_, 0, results.Size()),
-	}
-	nsIDBytes := ns.Bytes()
+	tagEncoder := s.pools.tagEncoder.Get()
```

Comment:
@robskillington does this make sense for getting a tag encoder?

Reply:
Yeah this is best way to do it and return at end of request.